<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SEESAW - Allelic expression analysis with Salmon and Swish • fishpond</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SEESAW - Allelic expression analysis with Salmon and Swish">
<meta property="og:description" content="fishpond">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fishpond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.3.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/swish.html">Swish: differential expression accounting for inferential uncertainty</a>
    </li>
    <li>
      <a href="../articles/allelic.html">SEESAW - Allelic expression analysis with Salmon and Swish</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mikelove/fishpond/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SEESAW - Allelic expression analysis with Salmon and Swish</h1>
            
            <h4 data-toc-skip class="date">08/02/2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mikelove/fishpond/blob/HEAD/vignettes/allelic.Rmd" class="external-link"><code>vignettes/allelic.Rmd</code></a></small>
      <div class="hidden name"><code>allelic.Rmd</code></div>

    </div>

    
    
<!-- run this document with rmarkdown::render("swish.Rmd") -->
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we describe usage of a suite of tools, <strong>SEESAW</strong>, Statistical Estimation of allelic Expression using Salmon and Swish, which allow for testing allelic imbalance across samples.</p>
<p><strong>Type of tests</strong></p>
<p>SEESAW allows for testing <em>global allelic imbalance</em> across all samples (pairwise testing within each individual), as well as <em>differential, or dynamic allelic imbalance</em> (pairwise allelic fold changes estimated within individual, followed by testing across two groups, or along an additional covariate). Each of these allelic imbalance (AI) analyses takes into account the potentially heterogeneous amount of inferential uncertainty per sample, per feature (transcript, transcript-group, or gene), and per allele.</p>
<p><strong>Steps in SEESAW</strong></p>
<p>Running SEESAW involves generation of a diploid transcriptome (e.g. using <a href="http://churchill-lab.github.io/g2gtools/" class="external-link">g2gtools</a>, construction of a diploid Salmon index (specifying <code>--keepDuplicates</code>), followed by Salmon quantification with a number of <a href="https://salmon.readthedocs.io/en/latest/salmon.html#numbootstraps" class="external-link">bootstrap inferential replicates</a> (we recommend 30 bootstrap replicates). These three steps (diploid reference preparation, indexing, quantification with bootstraps) provide the input data for the following statistical analyses in R/Bioconductor. The steps shown in this vignette leverage Bioconductor infrastructure including <em>SummarizedExperiment</em> for storage of input data and results, <em>tximport</em> for data import, and <em>GRanges</em> and <em>Gviz</em> for plotting.</p>
<p>In short the SEESAW steps are as listed, and diagrammed below:</p>
<ol style="list-style-type: decimal">
<li>g2gtools (diploid reference preparation)</li>
<li>Salmon indexing with <code>--keepDuplicates</code>
</li>
<li>Salmon quantification with bootstraps</li>
<li>
<code><a href="../reference/makeTx2Tss.html">makeTx2Tss()</a></code> aggregates data to TSS-level (optional)</li>
<li>
<code><a href="../reference/importAllelicCounts.html">importAllelicCounts()</a></code> creates a <em>SummarizedExperiment</em>
</li>
<li>Swish analysis: <code><a href="../reference/labelKeep.html">labelKeep()</a></code> and <code><a href="../reference/swish.html">swish()</a></code> (skip scaling)</li>
<li>Plotting</li>
</ol>
<p><img src="images/SEESAW.png" width="1240"></p>
<p>Below we demonstrate an analysis where transcripts are grouped by their transcription start site (TSS), although gene-level or transcript-level analysis is also possible. Additionally, any custom grouping could be used, by manually generating a <code>t2g</code> table as shown below. Special plotting functions in <em>fishpond</em> facilitate visualization of allelic and isoform changes at different resolutions, alongside gene models. In three examples, we perform global AI testing, differential AI testing, and dynamic AI testing, in all cases on simulated data associated with human genes.</p>
</div>
<div class="section level2">
<h2 id="linking-transcripts-to-tss">Linking transcripts to TSS<a class="anchor" aria-label="anchor" href="#linking-transcripts-to-tss"></a>
</h2>
<p>We begin assuming steps 1-3 have been completed. We can use the <code>makeTx2Tss</code> function to generate a <em>GRanges</em> object <code>t2g</code> that connects transcripts to transcript groups.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jorainer/ensembldb" class="external-link">ensembldb</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mikelove.github.io/fishpond">fishpond</a></span><span class="op">)</span></span>
<span><span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">t2g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeTx2Tss.html">makeTx2Tss</a></span><span class="op">(</span><span class="va">edb</span><span class="op">)</span> <span class="co"># GRanges object</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">t2g</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx_id"</span>,<span class="st">"group_id"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 216741 rows and 2 columns</span></span>
<span><span class="co">##                           tx_id               group_id</span></span>
<span><span class="co">##                     &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">## ENST00000456328 ENST00000456328  ENSG00000223972-11869</span></span>
<span><span class="co">## ENST00000450305 ENST00000450305  ENSG00000223972-12010</span></span>
<span><span class="co">## ENST00000488147 ENST00000488147  ENSG00000227232-29570</span></span>
<span><span class="co">## ENST00000619216 ENST00000619216  ENSG00000278267-17436</span></span>
<span><span class="co">## ENST00000473358 ENST00000473358  ENSG00000243485-29554</span></span>
<span><span class="co">## ...                         ...                    ...</span></span>
<span><span class="co">## ENST00000420810 ENST00000420810 ENSG00000224240-2654..</span></span>
<span><span class="co">## ENST00000456738 ENST00000456738 ENSG00000227629-2659..</span></span>
<span><span class="co">## ENST00000435945 ENST00000435945 ENSG00000237917-2663..</span></span>
<span><span class="co">## ENST00000435741 ENST00000435741 ENSG00000231514-2662..</span></span>
<span><span class="co">## ENST00000431853 ENST00000431853 ENSG00000235857-5685..</span></span></code></pre>
<p>Alternatively for gene-level analysis, one could either prepare a <code>t2g</code> data.frame with <code>tx_id</code> and <code>gene_id</code> columns, or a <code>t2g</code> <em>GRanges</em> object with a column <code>group_id</code> that is equal to <code>gene_id</code>.</p>
</div>
<div class="section level2">
<h2 id="importing-allelic-counts">Importing allelic counts<a class="anchor" aria-label="anchor" href="#importing-allelic-counts"></a>
</h2>
<p>Here we will use simulated data, but we can import allelic counts with the <code><a href="../reference/importAllelicCounts.html">importAllelicCounts()</a></code> function. It is best to read over the manual page for this function. For TSS-level analysis, the <code>t2g</code> <em>GRanges</em> generated above should be passed to the <code>tx2gene</code> argument. This will summarize transcript-level counts to the TSS level, and will attach <code>rowRanges</code> that provide the genomic locations of the grouped transcripts. Note that <code>importAllelicCounts</code> does not yet have the ability to automatically generate ranges based on sequence hashing (as occurs in <code>tximeta</code>).</p>
</div>
<div class="section level2">
<h2 id="filtering-features-with-no-information">Filtering features with no information<a class="anchor" aria-label="anchor" href="#filtering-features-with-no-information"></a>
</h2>
<p>Because we use <code>--keepDuplicates</code> in the step when we build the Salmon index, there will be a number of features in which there is no information about the allelic expression in the reads. We can find these features in bootstrap data by examining when the inferential replicates are nearly identical for the two alleles, as this is how the EM will split the reads. Removing these features avoids downstream problems during differential testing. Code for this filtering follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span></span>
<span><span class="va">rep1a1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"infRep1"</span><span class="op">)</span><span class="op">[</span>,<span class="va">y</span><span class="op">$</span><span class="va">allele</span> <span class="op">==</span> <span class="st">"a1"</span><span class="op">]</span></span>
<span><span class="va">rep1a2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"infRep1"</span><span class="op">)</span><span class="op">[</span>,<span class="va">y</span><span class="op">$</span><span class="va">allele</span> <span class="op">==</span> <span class="st">"a2"</span><span class="op">]</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">someInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">rep1a1</span> <span class="op">-</span> <span class="va">rep1a2</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">n</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">someInfo</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="global-allelic-imbalance">Global allelic imbalance<a class="anchor" aria-label="anchor" href="#global-allelic-imbalance"></a>
</h2>
<p>We begin by generating a simulated data object that resembles what one would obtain with <code><a href="../reference/importAllelicCounts.html">importAllelicCounts()</a></code>. The import function arranges the <code>a2</code> (non-effect) allelic counts first, followed by the <code>a1</code> (effect) allelic counts. Allelic ratios are calculated as <code>a1/a2</code>, which follows the notational standard in PLINK and other tools.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSimSwishData.html">makeSimSwishData</a></span><span class="op">(</span>allelic<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20 rows and 2 columns</span></span>
<span><span class="co">##          allele   sample</span></span>
<span><span class="co">##        &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">## s1-a2        a2  sample1</span></span>
<span><span class="co">## s2-a2        a2  sample2</span></span>
<span><span class="co">## s3-a2        a2  sample3</span></span>
<span><span class="co">## s4-a2        a2  sample4</span></span>
<span><span class="co">## s5-a2        a2  sample5</span></span>
<span><span class="co">## ...         ...      ...</span></span>
<span><span class="co">## s6-a1        a1 sample6 </span></span>
<span><span class="co">## s7-a1        a1 sample7 </span></span>
<span><span class="co">## s8-a1        a1 sample8 </span></span>
<span><span class="co">## s9-a1        a1 sample9 </span></span>
<span><span class="co">## s10-a1       a1 sample10</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">allele</span><span class="op">)</span> <span class="co"># a1/a2 allelic fold changes</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "a2" "a1"</span></span></code></pre>
<p>A hidden code chunk is used to add ranges from the <em>EnsDb</em> to the simulated dataset. For a real dataset, the ranges would be added either by <code>importAllelicCounts</code> (if using <code>tx2gene</code>) or could be added manually for transcript- or gene-level analysis, using the <code>rowRanges&lt;-</code> setter function. The ranges are only needed for the <code>plotAllelicGene</code> plotting function below.</p>
<pre><code>&lt;hidden code chunk&gt;</code></pre>
<p>We can already plot a heatmap of allelic ratios, before performing statistical testing. We can see in the first gene, <em>ADSS</em>, there appear to be two groups of transcripts with opposing allelic fold change. SEESAW makes use of <em>pheatmap</em> for plotting a heatmap of allelic ratios.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeInfRV.html">computeInfRV</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="co"># for posterior mean, variance</span></span>
<span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">==</span> <span class="va">gene</span></span>
<span><span class="fu"><a href="../reference/plotAllelicHeatmap.html">plotAllelicHeatmap</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="va">idx</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-7-1.png"><!-- --></p>
<p>The following two functions perform a Swish analysis, comparing the allelic counts within sample, while accounting for uncertainty in the assignment of the reads. The underlying test statistic is a Wilcoxon signed-rank statistic, which compares the two allele counts from each sample, so a paired analysis.</p>
<p><strong>Scaling:</strong> Note that we do not use <code>scaleInfReps</code> in the allelic pipeline. Because we compare the two alleles within samples, there is no need to perform scaling of the counts to adjust for sequencing depth. We simply import counts, filter low counts with <code>lableKeep</code> and then run the statistical testing with <code>swish</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"allele"</span>, pair<span class="op">=</span><span class="st">"sample"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plotting-results">Plotting results<a class="anchor" aria-label="anchor" href="#plotting-results"></a>
</h2>
<p>We can return to the heatmap, and now add q-values, etc. For details on adding metadata to a <em>pheatmap</em> plot object, see <code>?pheatmap</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>minusLogQ<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                  row.names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotAllelicHeatmap.html">plotAllelicHeatmap</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="va">idx</span>, annotation_row<span class="op">=</span><span class="va">dat</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-9-1.png"><!-- --></p>
<p>In order to visualize the inferential uncertainty, we can make use of <code><a href="../reference/plotInfReps.html">plotInfReps()</a></code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4.1</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"allele"</span>, cov<span class="op">=</span><span class="st">"sample"</span>, xaxis<span class="op">=</span><span class="cn">FALSE</span>, xlab<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="fl">2</span>, x<span class="op">=</span><span class="st">"allele"</span>, cov<span class="op">=</span><span class="st">"sample"</span>, xaxis<span class="op">=</span><span class="cn">FALSE</span>, xlab<span class="op">=</span><span class="st">""</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-10-1.png"><!-- --></p>
<p><strong>Plotting results in genomic context</strong></p>
<p>For analysis at the isoform or TSS-level, it may be useful to display results within a gene, relating the allelic differences to various gene features. SEESAW provides <code><a href="../reference/plotAllelicGene.html">plotAllelicGene()</a></code> in order to build visualization of Swish test statistics, allelic proportions, and isoform proportions, in a genomic context, making use of <em>Gviz</em>. Note that this function is not relevant for gene-level AI analysis. The first three arguments to <code><a href="../reference/plotAllelicGene.html">plotAllelicGene()</a></code> are the <em>SummarizedExperiment</em> object, the name of a gene (should match <code>gene_id</code> column), and a <em>TxDb</em> or <em>EnsDb</em> to use for plotting the gene model at the top. The statistics and proportions are then plotted at the first position of the feature (<code>start</code> for <code>+</code> features and <code>end</code> for <code>-</code> features).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotAllelicGene.html">plotAllelicGene</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">gene</span>, <span class="va">edb</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-11-1.png"><!-- --></p>
<p>You can also specify the gene using <code>symbol</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAllelicGene.html">plotAllelicGene</a></span><span class="op">(</span><span class="va">y</span>, symbol<span class="op">=</span><span class="st">"ADSS"</span>, db<span class="op">=</span><span class="va">edb</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-12-1.png"><!-- --></p>
<p>In the allelic proportion and isoform proportion tracks, a line is drawn through the mean proportion for a2 and a1 allele, and for the isoform proportion, across samples, at the start site for each transcript group. The line is meant only to help visualize the mean value as it may change across transcript groups, but the line has no meaning in the ranges in between features. That is, unlike continuous genomic features (methylation or accessibility), there is no meaning to the allelic proportion or isoform proportion outside of measured start sites of transcription.</p>
<p>We can further customize the plot, for example, changing the labels displayed on the gene model, and changing the labels for the alleles. An ideogram can be added with <code>ideogram=TRUE</code>, although this requires connecting to an external FTP site.</p>
<p>See <code>importAllelicGene()</code> manual page for more details.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAllelicGene.html">plotAllelicGene</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">gene</span>, <span class="va">edb</span>,</span>
<span>                transcriptAnnotation<span class="op">=</span><span class="st">"transcript"</span>,</span>
<span>                labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a2<span class="op">=</span><span class="st">"maternal"</span>,a1<span class="op">=</span><span class="st">"paternal"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-13-1.png"><!-- --></p>
<p>We can also customize the display of the alleles in the <code><a href="../reference/plotInfReps.html">plotInfReps()</a></code> plots, by adding a new factor, while carefully noting the existing and new allele labels, to make sure the annotation is correct:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span><span class="op">$</span><span class="va">allele_new</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">allele</span></span>
<span><span class="co"># note a2 is non-effect, a1 is effect:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">allele</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "a2" "a1"</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># replace a2 then a1:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">allele_new</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"maternal"</span>,<span class="st">"paternal"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"allele_new"</span>, </span>
<span>            legend<span class="op">=</span><span class="cn">TRUE</span>, legendPos<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-14-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="differential-allelic-imbalance">Differential allelic imbalance<a class="anchor" aria-label="anchor" href="#differential-allelic-imbalance"></a>
</h2>
<p>Above, we tested for global AI, where the allelic fold change is consistent across all samples. We can also test for differential or dynamic AI, by adding specification of a <code>cov</code> (covariate) which can be either a two-group factor, or a continuous variable. Here we demonstrate differential AI, when <code>cov</code> is a two-group factor, in this case called <code>"condition"</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSimSwishData.html">makeSimSwishData</a></span><span class="op">(</span>diffAI<span class="op">=</span><span class="cn">TRUE</span>, n<span class="op">=</span><span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 24 rows and 3 columns</span></span>
<span><span class="co">##          allele   sample condition</span></span>
<span><span class="co">##        &lt;factor&gt; &lt;factor&gt;  &lt;factor&gt;</span></span>
<span><span class="co">## s1-a2        a2  sample1         A</span></span>
<span><span class="co">## s2-a2        a2  sample2         A</span></span>
<span><span class="co">## s3-a2        a2  sample3         A</span></span>
<span><span class="co">## s4-a2        a2  sample4         A</span></span>
<span><span class="co">## s5-a2        a2  sample5         A</span></span>
<span><span class="co">## ...         ...      ...       ...</span></span>
<span><span class="co">## s8-a1        a1 sample8          B</span></span>
<span><span class="co">## s9-a1        a1 sample9          B</span></span>
<span><span class="co">## s10-a1       a1 sample10         B</span></span>
<span><span class="co">## s11-a1       a1 sample11         B</span></span>
<span><span class="co">## s12-a1       a1 sample12         B</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">condition</span>, <span class="va">y</span><span class="op">$</span><span class="va">allele</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    </span></span>
<span><span class="co">##     a2 a1</span></span>
<span><span class="co">##   A  6  6</span></span>
<span><span class="co">##   B  6  6</span></span></code></pre>
<p>In the following, we test for changes in allelic imbalance across <code>condition</code>. This is implemented as an “interaction” test, where we test if the fold change associated with <code>allele</code>, for paired samples, differs across <code>condition</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"allele"</span>, pair<span class="op">=</span><span class="st">"sample"</span>,</span>
<span>           cov<span class="op">=</span><span class="st">"condition"</span>, interaction<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>In this simulated data, the top two features exhibit differential AI with low uncertainty, so these emerge as highly significant, as expected.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stat"</span>,<span class="st">"qvalue"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">##             stat    qvalue</span></span>
<span><span class="co">##        &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## gene-1      17.3     0.005</span></span>
<span><span class="co">## gene-2     -17.5     0.005</span></span></code></pre>
<p>The non-AI features have roughly uniform p-values:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,<span class="st">"pvalue"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-18-1.png"><!-- --></p>
<p>We can plot the allelic counts with uncertainty, grouped by the condition (black and grey lines at bottom).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"allele"</span>, cov<span class="op">=</span><span class="st">"condition"</span>,</span>
<span>            xaxis<span class="op">=</span><span class="cn">FALSE</span>, legend<span class="op">=</span><span class="cn">TRUE</span>, legendPos<span class="op">=</span><span class="st">"bottomright"</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-19-1.png"><!-- --></p>
<p>We can also visualize the data across multiple features, in terms of allelic ratios:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">row_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>minusLogQ<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                      row.names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">col_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>condition<span class="op">=</span><span class="va">y</span><span class="op">$</span><span class="va">condition</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>                      row.names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"s"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotAllelicHeatmap.html">plotAllelicHeatmap</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="va">idx</span>,</span>
<span>                   annotation_row<span class="op">=</span><span class="va">row_dat</span>,</span>
<span>                   annotation_col<span class="op">=</span><span class="va">col_dat</span>,</span>
<span>                   cluster_rows<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-20-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="dynamic-allelic-imbalance">Dynamic allelic imbalance<a class="anchor" aria-label="anchor" href="#dynamic-allelic-imbalance"></a>
</h2>
<p>Now we demonstrate dynamic AI testing when <code>cov</code> (covariate) is a continuous variable. In this case, the user should specify a correlation test, either <code>cor="pearson"</code> or <code>"spearman"</code>, which is the underlying test statistic used by Swish (it will then be averaged over bootstraps and a permutation null is generated to assess FDR). We have found that Pearson correlations work well in our testing, but the Spearman correlation offers additional robustness against outlying values in <code>cov</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSimSwishData.html">makeSimSwishData</a></span><span class="op">(</span>dynamicAI<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20 rows and 3 columns</span></span>
<span><span class="co">##          allele   sample      time</span></span>
<span><span class="co">##        &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## s1-a2        a2  sample1      0.00</span></span>
<span><span class="co">## s2-a2        a2  sample2      0.11</span></span>
<span><span class="co">## s3-a2        a2  sample3      0.22</span></span>
<span><span class="co">## s4-a2        a2  sample4      0.33</span></span>
<span><span class="co">## s5-a2        a2  sample5      0.44</span></span>
<span><span class="co">## ...         ...      ...       ...</span></span>
<span><span class="co">## s6-a1        a1 sample6       0.56</span></span>
<span><span class="co">## s7-a1        a1 sample7       0.67</span></span>
<span><span class="co">## s8-a1        a1 sample8       0.78</span></span>
<span><span class="co">## s9-a1        a1 sample9       0.89</span></span>
<span><span class="co">## s10-a1       a1 sample10      1.00</span></span></code></pre>
<p>A hidden code chunk adds ranges to our simulation data.</p>
<pre><code>&lt;hidden code chunk&gt;</code></pre>
<p>In the following, we test for changes in allelic imbalance within sample that correlate with a covariate <code>time</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"allele"</span>, pair<span class="op">=</span><span class="st">"sample"</span>, cov<span class="op">=</span><span class="st">"time"</span>, cor<span class="op">=</span><span class="st">"pearson"</span><span class="op">)</span></span></code></pre></div>
<p>Note the first two features have small q-values and opposite test statistic; here the test statistic is the average Pearson correlation of the allelic log fold change with the <code>time</code> variable, averaging over bootstrap replicates.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stat"</span>,<span class="st">"qvalue"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">##                     stat    qvalue</span></span>
<span><span class="co">##                &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## ADSS-244452134  0.870969     0.005</span></span>
<span><span class="co">## ADSS-244419273 -0.861573     0.005</span></span></code></pre>
<p>For plotting inferential replicates over a continuous variable, we must first compute summary statistics of inferential mean and variance:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeInfRV.html">computeInfRV</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p>Now we can examine the allelic counts across the <code>time</code> variable:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2.5</span>,<span class="fl">4</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"time"</span>, cov<span class="op">=</span><span class="st">"allele"</span>, shiftX<span class="op">=</span><span class="fl">.01</span>, xaxis<span class="op">=</span><span class="cn">FALSE</span>, xlab<span class="op">=</span><span class="st">""</span>, main<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4.5</span>,<span class="fl">4</span>,<span class="fl">0</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="fl">2</span>, x<span class="op">=</span><span class="st">"time"</span>, cov<span class="op">=</span><span class="st">"allele"</span>, shiftX<span class="op">=</span><span class="fl">.01</span>, main<span class="op">=</span><span class="st">""</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-26-1.png"><!-- --></p>
<p>With a little more code, we can add a <code>lowess</code> line for each series:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="fl">1</span>, x<span class="op">=</span><span class="st">"time"</span>, cov<span class="op">=</span><span class="st">"allele"</span>, shiftX<span class="op">=</span><span class="fl">.01</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  time <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">time</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,</span>
<span>  a2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"mean"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>,<span class="va">y</span><span class="op">$</span><span class="va">allele</span><span class="op">==</span><span class="st">"a2"</span><span class="op">]</span>,</span>
<span>  a1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"mean"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>,<span class="va">y</span><span class="op">$</span><span class="va">allele</span><span class="op">==</span><span class="st">"a1"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lowess.html" class="external-link">lowess</a></span><span class="op">(</span><span class="va">dat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"dodgerblue"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lowess.html" class="external-link">lowess</a></span><span class="op">(</span><span class="va">dat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"goldenrod4"</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-27-1.png"><!-- --></p>
<p>Visualizing the allelic proportion in a heatmap helps to see relationships with the <code>time</code> variable, while also showing data from multiple features at once:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">row_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>minusLogQ<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                      row.names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">col_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>time<span class="op">=</span><span class="va">y</span><span class="op">$</span><span class="va">time</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,</span>
<span>                      row.names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"s"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotAllelicHeatmap.html">plotAllelicHeatmap</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="va">idx</span>,</span>
<span>                   annotation_row<span class="op">=</span><span class="va">row_dat</span>,</span>
<span>                   annotation_col<span class="op">=</span><span class="va">col_dat</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-28-1.png"><!-- --></p>
<p><strong>Plotting results in genomic context</strong></p>
<p>Previously, in the global AI section, we demonstrated how to plot TSS-level results in a genomic context using <code><a href="../reference/plotAllelicGene.html">plotAllelicGene()</a></code>. Here we demonstrate how to repeat such a plot for differential or dynamic AI analysis. There is an extra step for dynamic analysis (binning the continuous covariate into groups) but otherwise the code would be similar.</p>
<p>We begin by binning the <code>time</code> covariate into a few groups, so that we can diagram the allelic and isoform proportions in the genomic context, but facetting across time.</p>
<p>We create the binned covariate using <code>cut</code>, and rename the labels for nicer labels in our plot. For differential AI, this step would be skipped (as there already exists a two-group covariate for grouping samples).</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span><span class="op">$</span><span class="va">time_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">time</span>,breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">.25</span>,<span class="fl">.75</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                   include.lowest<span class="op">=</span><span class="cn">TRUE</span>, labels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">y</span><span class="op">$</span><span class="va">time_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"time-"</span>,<span class="va">y</span><span class="op">$</span><span class="va">time_bins</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">time_bins</span><span class="op">[</span> <span class="va">y</span><span class="op">$</span><span class="va">allele</span> <span class="op">==</span> <span class="st">"a2"</span> <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## time-1 time-2 time-3 </span></span>
<span><span class="co">##      3      4      3</span></span></code></pre>
<p>We can then make our facetted allelic proportion plot:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotAllelicGene.html">plotAllelicGene</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">gene</span>, <span class="va">edb</span>, cov<span class="op">=</span><span class="st">"time_bins"</span>,</span>
<span>                qvalue<span class="op">=</span><span class="cn">FALSE</span>, log2FC<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-30-1.png"><!-- --></p>
<p>If we also want to visualize how isoform proportions may be changing, we can add <code>covFacetIsoform=TRUE</code>, which additionally facets the isoform proportion plot by the covariate:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAllelicGene.html">plotAllelicGene</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">gene</span>, <span class="va">edb</span>, cov<span class="op">=</span><span class="st">"time_bins"</span>,</span>
<span>                covFacetIsoform<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                qvalue<span class="op">=</span><span class="cn">FALSE</span>, log2FC<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_files/figure-html/unnamed-chunk-31-1.png"><!-- --></p>
<p>For further questions about the SEESAW steps, please post to one of these locations:</p>
<ul>
<li>Bioconductor support site <a href="https://support.bioconductor.org" class="external-link uri">https://support.bioconductor.org</a> and use the tag <code>fishpond</code> or <code>swish</code>
</li>
<li>GitHub Issue <a href="https://github.com/mikelove/fishpond" class="external-link uri">https://github.com/mikelove/fishpond</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.1.1 (2021-08-10)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SummarizedExperiment_1.24.0 MatrixGenerics_1.6.0       </span></span>
<span><span class="co">##  [3] matrixStats_0.62.0          fishpond_2.3.10            </span></span>
<span><span class="co">##  [5] EnsDb.Hsapiens.v86_2.99.0   ensembldb_2.18.4           </span></span>
<span><span class="co">##  [7] AnnotationFilter_1.18.0     GenomicFeatures_1.46.5     </span></span>
<span><span class="co">##  [9] AnnotationDbi_1.56.2        Biobase_2.54.0             </span></span>
<span><span class="co">## [11] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        </span></span>
<span><span class="co">## [13] IRanges_2.28.0              S4Vectors_0.32.4           </span></span>
<span><span class="co">## [15] BiocGenerics_0.40.0        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] backports_1.4.1             Hmisc_4.7-0                </span></span>
<span><span class="co">##   [3] BiocFileCache_2.2.1         systemfonts_1.0.4          </span></span>
<span><span class="co">##   [5] plyr_1.8.7                  lazyeval_0.2.2             </span></span>
<span><span class="co">##   [7] splines_4.1.1               BiocParallel_1.28.3        </span></span>
<span><span class="co">##   [9] ggplot2_3.3.6               digest_0.6.29              </span></span>
<span><span class="co">##  [11] htmltools_0.5.3             fansi_1.0.3                </span></span>
<span><span class="co">##  [13] magrittr_2.0.3              checkmate_2.1.0            </span></span>
<span><span class="co">##  [15] memoise_2.0.1               svMisc_1.2.3               </span></span>
<span><span class="co">##  [17] BSgenome_1.62.0             cluster_2.1.3              </span></span>
<span><span class="co">##  [19] Biostrings_2.62.0           pkgdown_2.0.6              </span></span>
<span><span class="co">##  [21] prettyunits_1.1.1           jpeg_0.1-9                 </span></span>
<span><span class="co">##  [23] colorspace_2.0-3            blob_1.2.3                 </span></span>
<span><span class="co">##  [25] rappdirs_0.3.3              textshaping_0.3.6          </span></span>
<span><span class="co">##  [27] xfun_0.31                   dplyr_1.0.9                </span></span>
<span><span class="co">##  [29] crayon_1.5.1                RCurl_1.98-1.8             </span></span>
<span><span class="co">##  [31] jsonlite_1.8.0              survival_3.3-1             </span></span>
<span><span class="co">##  [33] VariantAnnotation_1.40.0    glue_1.6.2                 </span></span>
<span><span class="co">##  [35] gtable_0.3.0                zlibbioc_1.40.0            </span></span>
<span><span class="co">##  [37] XVector_0.34.0              DelayedArray_0.20.0        </span></span>
<span><span class="co">##  [39] SingleCellExperiment_1.16.0 abind_1.4-5                </span></span>
<span><span class="co">##  [41] scales_1.2.0                pheatmap_1.0.12            </span></span>
<span><span class="co">##  [43] DBI_1.1.3                   Rcpp_1.0.9                 </span></span>
<span><span class="co">##  [45] progress_1.2.2              htmlTable_2.4.1            </span></span>
<span><span class="co">##  [47] foreign_0.8-82              bit_4.0.4                  </span></span>
<span><span class="co">##  [49] Formula_1.2-4               htmlwidgets_1.5.4          </span></span>
<span><span class="co">##  [51] httr_1.4.3                  RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [53] ellipsis_0.3.2              pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [55] XML_3.99-0.10               farver_2.1.1               </span></span>
<span><span class="co">##  [57] Gviz_1.38.4                 nnet_7.3-17                </span></span>
<span><span class="co">##  [59] sass_0.4.2                  dbplyr_2.2.1               </span></span>
<span><span class="co">##  [61] deldir_1.0-6                utf8_1.2.2                 </span></span>
<span><span class="co">##  [63] tidyselect_1.1.2            rlang_1.0.4                </span></span>
<span><span class="co">##  [65] reshape2_1.4.4              munsell_0.5.0              </span></span>
<span><span class="co">##  [67] tools_4.1.1                 cachem_1.0.6               </span></span>
<span><span class="co">##  [69] cli_3.3.0                   generics_0.1.3             </span></span>
<span><span class="co">##  [71] RSQLite_2.2.15              evaluate_0.15              </span></span>
<span><span class="co">##  [73] stringr_1.4.0               fastmap_1.1.0              </span></span>
<span><span class="co">##  [75] yaml_2.3.5                  ragg_1.2.2                 </span></span>
<span><span class="co">##  [77] knitr_1.39                  bit64_4.0.5                </span></span>
<span><span class="co">##  [79] fs_1.5.2                    purrr_0.3.4                </span></span>
<span><span class="co">##  [81] KEGGREST_1.34.0             xml2_1.3.3                 </span></span>
<span><span class="co">##  [83] biomaRt_2.50.3              compiler_4.1.1             </span></span>
<span><span class="co">##  [85] rstudioapi_0.13             filelock_1.0.2             </span></span>
<span><span class="co">##  [87] curl_4.3.2                  png_0.1-7                  </span></span>
<span><span class="co">##  [89] tibble_3.1.8                bslib_0.4.0                </span></span>
<span><span class="co">##  [91] stringi_1.7.8               highr_0.9                  </span></span>
<span><span class="co">##  [93] desc_1.4.1                  lattice_0.20-45            </span></span>
<span><span class="co">##  [95] ProtGenerics_1.26.0         Matrix_1.4-1               </span></span>
<span><span class="co">##  [97] vctrs_0.4.1                 pillar_1.8.0               </span></span>
<span><span class="co">##  [99] lifecycle_1.0.1             jquerylib_0.1.4            </span></span>
<span><span class="co">## [101] data.table_1.14.2           bitops_1.0-7               </span></span>
<span><span class="co">## [103] rtracklayer_1.54.0          qvalue_2.26.0              </span></span>
<span><span class="co">## [105] R6_2.5.1                    BiocIO_1.4.0               </span></span>
<span><span class="co">## [107] latticeExtra_0.6-30         gridExtra_2.3              </span></span>
<span><span class="co">## [109] dichromat_2.0-0.1           gtools_3.9.3               </span></span>
<span><span class="co">## [111] assertthat_0.2.1            rprojroot_2.0.3            </span></span>
<span><span class="co">## [113] rjson_0.2.21                GenomicAlignments_1.30.0   </span></span>
<span><span class="co">## [115] Rsamtools_2.10.0            GenomeInfoDbData_1.2.7     </span></span>
<span><span class="co">## [117] parallel_4.1.1              hms_1.1.1                  </span></span>
<span><span class="co">## [119] grid_4.1.1                  rpart_4.1.16               </span></span>
<span><span class="co">## [121] rmarkdown_2.14              biovizBase_1.42.0          </span></span>
<span><span class="co">## [123] base64enc_0.1-3             interp_1.1-3               </span></span>
<span><span class="co">## [125] restfulr_0.0.15</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/aqzhu91" class="external-link">Anqi Zhu</a>, <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://k3yavi.github.io" class="external-link">Avi Srivastava</a>, <a href="https://combine-lab.github.io" class="external-link">Rob Patro</a>, <a href="https://sph.unc.edu/adv_profile/joseph-g-ibrahim-phd" class="external-link">Joseph Ibrahim</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
